python_modules/dagster/dagster_tests/definitions_tests/test_materialize_result.py

```diff
@@ -0,0 +1,350 @@
+                metadata={"one": 1},
+            ),
+            MaterializeResult(
+                asset_key="two",
+                metadata={"two": 2},
+            ),
+        )
+
+    mats = _exec_asset(ret_multi)
+
+    assert len(mats) == 2, mats
+    assert "one" in mats[0].metadata
+    assert mats[0].tags
+    assert "two" in mats[1].metadata
+    assert mats[1].tags
+
+    #
+    # return list error
+    #
+    @multi_asset(outs={"one": AssetOut(), "two": AssetOut()})
+    def ret_list():
+        return [
+            MaterializeResult(
+                asset_key="one",
+                metadata={"one": 1},
+            ),
+            MaterializeResult(
+                asset_key="two",
+                metadata={"two": 2},
+            ),
+        ]
+
+    # not the best
+    with pytest.raises(
+        DagsterInvariantViolationError,
+        match=(
+            "When using multiple outputs, either yield each output, or return a tuple containing a"
+            " value for each output."
+        ),
+    ):
+        _exec_asset(ret_list)
+
+
+def test_materialize_result_output_typing():
+    # Test that the return annotation MaterializeResult is interpreted as a Nothing type, since we
+    # coerce returned MaterializeResults to Output(None)
+
+    class TestingIOManager(IOManager):
+        def handle_output(self, context, obj):
+            assert context.dagster_type.is_nothing
+            return None
+
+        def load_input(self, context):
+            return 1
+
+    @asset
+    def asset_with_type_annotation() -> MaterializeResult:
+        return MaterializeResult(metadata={"foo": "bar"})
+
+    assert materialize(
+        [asset_with_type_annotation], resources={"io_manager": TestingIOManager()}
+    ).success
+
+    @multi_asset(outs={"one": AssetOut(), "two": AssetOut()})
+    def multi_asset_with_outs_and_type_annotation() -> Tuple[MaterializeResult, MaterializeResult]:
+        return MaterializeResult(asset_key="one"), MaterializeResult(asset_key="two")
+
+    assert materialize(
+        [multi_asset_with_outs_and_type_annotation], resources={"io_manager": TestingIOManager()}
+    ).success
+
+    @multi_asset(specs=[AssetSpec("one"), AssetSpec("two")])
+    def multi_asset_with_specs_and_type_annotation() -> Tuple[MaterializeResult, MaterializeResult]:
+        return MaterializeResult(asset_key="one"), MaterializeResult(asset_key="two")
+
+    assert materialize(
+        [multi_asset_with_specs_and_type_annotation], resources={"io_manager": TestingIOManager()}
+    ).success
+
+    @multi_asset(specs=[AssetSpec("one"), AssetSpec("two")])
+    def multi_asset_with_specs_and_no_type_annotation():
+        return MaterializeResult(asset_key="one"), MaterializeResult(asset_key="two")
+
+    assert materialize(
+        [multi_asset_with_specs_and_no_type_annotation],
+        resources={"io_manager": TestingIOManager()},
+    ).success
+
+
+@pytest.mark.skip("generator return types are interpreted as Any")
```

Comment:
marked as skipped because Generator types are converted to `Any`. This also means that `MaterializeResult`s will be stored by the I/O manager. But handling generator type hints opens up a big can of worms that will probably be addressed by https://github.com/dagster-io/dagster/pull/16906

Reply:
i tend to think it's fine to ship without fixing this issue, as it sounds like predates to MaterializeResult - as opposed to MaterializeResult specific, it seems to be a bigger issue in the framework as mentioned in nick's RFC.